src/app/clusters/im-compatibility/ember-compatibility-functions.cpp

```diff
@@ -0,0 +1,89 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *          Contains the functions for compatibility with ember ZCL inner state
+ *          when calling ember callbacks.
+ */
+
+#include "ember-compatibility-functions.h"
+
+#include <lib/core/CHIPCore.h>
+#include <lib/core/CHIPTLV.h>
+#include <lib/support/CodeUtils.h>
+#include <util/util.h>
+
+namespace chip {
+namespace app {
+namespace Compatibility {
+namespace {
+EmberAfClusterCommand imCompatibilityEmberAfCluster;
+EmberApsFrame imCompatibilityEmberApsFrame;
+EmberAfInterpanHeader imCompatibilityInterpanHeader;
+Command * currentCommandObject;
+} // namespace
+
+void SetupEmberAfObjects(Command * command, ClusterId clusterId, CommandId commandId, EndpointId endpointId)
+{
+    imCompatibilityEmberApsFrame.clusterId           = clusterId;
+    imCompatibilityEmberApsFrame.destinationEndpoint = endpointId;
+    imCompatibilityEmberApsFrame.sourceEndpoint      = endpointId;
+    imCompatibilityEmberAfCluster.commandId          = commandId;
```

Comment:
Can you fill the `source` with the `nodeId` of the source here and similarly to `EmberApsFrame` can you fill `seqNum` with some bits of the exchange id ?

Reply:
In theory, we should also fill `mfgCode` and `mfgSpecific`. Not sure the IM has anything implemented for them yet ?

Reply:
I did not found fields that is related to mfgCode in the CHIP spec, not sure if this is intentional.

Reply:
> I did not found fields that is related to mfgCode in the CHIP spec, not sure if this is intentional or maybe I missed some words in the spec?

I have not found anything either in the spec.
One example of how it is used in the current code is to differentiate between commands. For example, the command id of `MfgSpecificPing` is `0x0000` which is the same than the command id of `ResetToFactoryDefaults`.
The `mfgCode` and `mfgSpecific` bits are then used to differentiate between those commands.
@bzbarsky-apple Are you aware of anything spec'ed for manufacturer specific extensions ?